### PR TITLE
React dynamic list of todos

### PR DESCRIPTION
**Pull request by vsupruniuk**
[DEMO LINK](https://vsupruniuk.github.io/react_dynamic-list-of-todos/)

### CODE REVIEW

File: src/components/TodoList/TodoList.tsx [hunk 1]
```diff
@@ -1,44 +1,149 @@
...
-            User&nbsp;#2
-          </button>
-        </li>
-      </ul>
+type Props = {
+  onUserIdChange: (userId: number) => void,
+};
+
+export const TodoList: React.FC<Props> = ({ onUserIdChange }) => {
+  const [todos, setTodos] = useState<Todo[] | []>([]);
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:13]: suggestion: if you add `...| []` just to make it a default value, you'd rather implement it the following way
```suggestion
  const [todos, setTodos] = useState<Todo[] | null>(null);
```

File: src/components/TodoList/TodoList.tsx [hunk 2]
```diff
@@ -1,44 +1,149 @@
...
+
+  const handleQueryChange = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setQuery(event.target.value);
+  };
+
+  const handleFilterChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setFilterByStatus(event.target.value);
+  };
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:29]: apply `useCallback` to avoid extra re-rendering

File: src/components/TodoList/TodoList.tsx [hunk 3]
```diff
@@ -1,44 +1,149 @@
...
+  //     currentIndex -= 1;
+
+  //     [randomizerArray[currentIndex], randomizerArray[randomIndex]]
+  //     = [randomizerArray[randomIndex], randomizerArray[currentIndex]];
+  //   }
+
+  //   return randomizerArray;
+  // };
+
+  let visibleTodos = todos;
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:47]: don't leave such variables this way, either wrap them or move them to state

File: src/components/TodoList/TodoList.tsx [hunk 4]
```diff
@@ -1,44 +1,149 @@
...
+                  TodoList__user-button
+                  TodoList__user-button--selected
+                  button
+                "
+                type="button"
+                onClick={() => {
+                  if (todo.userId) {
+                    onUserIdChange(todo.userId);
+                  }
+                }}
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:139]: create a separate function

File: src/components/TodoList/TodoList.tsx [hunk 5]
```diff
@@ -1,44 +1,149 @@
...
-          >
-            User&nbsp;#2
-          </button>
-        </li>
-      </ul>
+type Props = {
+  onUserIdChange: (userId: number) => void,
+};
+
+export const TodoList: React.FC<Props> = ({ onUserIdChange }) => {
```
Comment (jstmpelowycz) [inline src/components/TodoList/TodoList.tsx:12]: `memo`?

File: src/components/CurrentUser/CurrentUser.tsx [hunk 6]
```diff
@@ -1,12 +1,46 @@
...
-export const CurrentUser: React.FC = () => (
-  <div className="CurrentUser">
-    <h2 className="CurrentUser__title"><span>Selected user: 2</span></h2>
+type Props = {
+  userId: number,
+  onUserIdChange: (userId: number) => void,
+};
+
+export const CurrentUser: React.FC<Props> = ({ userId, onUserIdChange }) => {
+  const [selectedUser, setSelectedUser] = useState<User | null>(null);
```
[the remark this answers is not in the document]
Comment (jstmpelowycz) [reply src/components/CurrentUser/CurrentUser.tsx:13]: also,`memo`?

**Review by jstmpelowycz — COMMENTED** [review]
good job
but – you've already learned more about advanced hooks, thus consider applying them there needed

**Review by jstmpelowycz — CHANGES_REQUESTED** [review]
needs a few fixes

File: src/components/TodoList/TodoList.tsx [hunk 7]
```diff
@@ -1,44 +1,168 @@
...
+                  button
+                "
+                type="button"
+                onClick={() => {
+                  handleUserIdChange(todo);
+                }}
+              >
+                {`User #${todo.userId}`}
+              </button>
+            </li>
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:162]: `TodoInfo` component?

File: src/components/TodoList/TodoList.tsx [hunk 8]
```diff
@@ -1,44 +1,168 @@
...
+    } catch (error) {
+      throw new Error('Failed to load todos from server');
+    }
+  }, []);
+
+  const onFilteringByQuery = () => {
+    setVisibleTodos([...visibleTodos].filter(
+      todo => todo.title.toLowerCase().includes(query.toLowerCase()),
+    ));
+  };
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:37]: move this to `handleQueryChange`, you don't need 2 calls in JSX handler

File: src/components/TodoList/TodoList.tsx [hunk 9]
```diff
@@ -1,44 +1,168 @@
...
+        break;
+
+      case 'completed':
+        setVisibleTodos([...todos].filter(todo => todo.completed));
+        break;
+
+      default:
+        setVisibleTodos([...todos]);
+    }
+  };
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:52]: Move this to handleFilterChange

File: src/components/TodoList/TodoList.tsx [hunk 10]
```diff
@@ -1,44 +1,168 @@
...
+
+  const onFilteringByQuery = () => {
+    setVisibleTodos([...visibleTodos].filter(
+      todo => todo.title.toLowerCase().includes(query.toLowerCase()),
+    ));
+  };
+
+  const onFilteringByStatus = (sortBy: string) => {
+    switch (sortBy) {
+      case 'active':
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:41]: Use typescript, create `enum` for these values

File: src/components/TodoList/TodoList.tsx [hunk 11]
```diff
@@ -1,44 +1,168 @@
...
+          aria-label="Default select example"
+          value={filterByStatus}
+          onChange={(event) => {
+            handleFilterChange(event);
+            onFilteringByStatus(event.target.value);
+          }}
+        >
+          <option value="all">All</option>
+          <option value="active">Active</option>
+          <option value="completed">Completed</option>
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:116]: So you can map your enum here

**Review by illiarerun — CHANGES_REQUESTED** [review]
Your search doesn't work
image
Randomize button also doesn't work, try to deploy again

File: src/components/TodoList/TodoList.tsx (hunk 8 above)
Comment (vsupruniuk) [reply src/components/TodoList/TodoList.tsx]: This approach causes problems. We try to filter the array by the new state before it is updated. So I combined both filtering methods inside a `useEffect `hook

File: src/components/TodoList/TodoList.tsx (hunk 9 above)
Comment (vsupruniuk) [reply src/components/TodoList/TodoList.tsx]: This approach causes problems. We try to filter the array by the new state before it is updated. So I combined both filtering methods inside a `useEffect `hook

File: src/components/TodoInfo/TodoInfo.tsx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  handleUserIdChange: (todo: Todo) => void,
+};
+
+export const TodoInfo: React.FC<Props> = React.memo(({
+  visibleTodos,
+  handleUserIdChange,
+}) => {
+  return (
+    <ul data-cy="listOfTodos" className="TodoList__list">
+      {visibleTodos?.map(todo => (
```
Comment (illiarerun) [inline src/components/TodoInfo/TodoInfo.tsx:17]: No this should be in `TodoList`  component

File: src/components/TodoList/TodoList.tsx [hunk 13]
```diff
@@ -1,44 +1,147 @@
...
+          className="form-select"
+          aria-label="Default select example"
+          value={filterByStatus}
+          onChange={(event) => {
+            handleFilterChange(event);
+          }}
+        >
+          <option value={SortType.ALL}>All</option>
+          <option value={SortType.ACTIVE}>Active</option>
+          <option value={SortType.COMPLETED}>Completed</option>
```
Comment (illiarerun) [inline src/components/TodoList/TodoList.tsx:128]: Map it using `Object.keys`

**Review by illiarerun — APPROVED** [review]

**Pull request closed without merging**